templates/grid_broker/grid_broker.py

```diff
@@ -45,53 +48,99 @@ def _watch_transactions(self):
                 continue
             # try to parse the transaction data
             try:
+                malformed = True
                 data = self._parse_tx_data(tx)
+                malformed = False
                 # refund if there is not data
                 if not data:
+                    self.logger.info("no data found in transaction %s", tx.id)
+                    raise ValueError("transaction has no data")
```

Comment:
This is not right. Here you raise cause there is no data.
So you end up in the except block line 58. And in this block you try to refund. But refund will always fail if there is not data cause you don't know the email address of the sender.

Reply:
Also I don't really understand why you need the `malformed` variable. The only thing that happens in this try block is the `_parse_tx_data`.  So the only case where you can end up in the except block is if the parsing of the transaction failed.

Reply:
I think we need to split this more, something like:
```python
data = None
try:
   data = parse_tx(data)
except:
   log.error

if not data:
    log error
     return

try:
    deploy
except:
    refund
```

Reply:
First of all, I didn't change the logic at all, I only restructured it so we don't have a refund block twice and hardened it so there is a try/except around the refund. It already did used to refund if there was no data. I don't understand the relationship between data being empty and not being able to refund. This is the data from the notary, the refund function doesn't use the data anyway. Also, if we can attempt to refund when the data can't be parsed, then we can definitely attempt to refund when the data is empty because when parsing fails, we also end up with no data. I just raise a value error so I can take advantage of the refund in the try/except instead of having duplicate code just two lines above it. The `malformed` variable is just there for sake of logging so I can differentiate between both cases.

Reply:
yeah I got confused, the data is indeed not needed to refund, just to notify the user via email.